flang/lib/Lower/Allocatable.cpp

```diff
@@ -0,0 +1,393 @@
+                                         upperBound};
+  llvm::SmallVector<mlir::Value, 4> operands;
+  for (auto [fst, snd] : llvm::zip(args, callee.getType().getInputs()))
+    operands.emplace_back(builder.createConvert(loc, snd, fst));
+  builder.create<fir::CallOp>(loc, callee, operands);
+}
+
+static mlir::Value
+genAllocatableAllocate(Fortran::lower::FirOpBuilder &builder,
+                       mlir::Location loc, mlir::Value boxAddress,
+                       mlir::Value hasStat, mlir::Value errMsgBox,
+                       mlir::Value sourceFile, mlir::Value sourceLine) {
+  auto callee =
+      Fortran::lower::genRuntimeFunction<mkRTKey(AllocatableAllocate)>(loc,
+                                                                       builder);
+  llvm::SmallVector<mlir::Value, 5> args{boxAddress, hasStat, errMsgBox,
+                                         sourceFile, sourceLine};
+  llvm::SmallVector<mlir::Value, 5> operands;
+  for (auto [fst, snd] : llvm::zip(args, callee.getType().getInputs()))
+    operands.emplace_back(builder.createConvert(loc, snd, fst));
+  return builder.create<fir::CallOp>(loc, callee, operands).getResult(0);
+}
+
+static mlir::Value
+genAllocatableDeallocate(Fortran::lower::FirOpBuilder &builder,
+                         mlir::Location loc, mlir::Value boxAddress,
+                         mlir::Value hasStat, mlir::Value errMsgBox,
+                         mlir::Value sourceFile, mlir::Value sourceLine) {
+  auto callee =
+      Fortran::lower::genRuntimeFunction<mkRTKey(AllocatableDeallocate)>(
+          loc, builder);
+  llvm::SmallVector<mlir::Value, 5> args{boxAddress, hasStat, errMsgBox,
+                                         sourceFile, sourceLine};
+  llvm::SmallVector<mlir::Value, 5> operands;
+  for (auto [fst, snd] : llvm::zip(args, callee.getType().getInputs()))
+    operands.emplace_back(builder.createConvert(loc, snd, fst));
+  return builder.create<fir::CallOp>(loc, callee, operands).getResult(0);
+}
+
+/// Helper to get symbol from AllocateObject.
+static const Fortran::semantics::Symbol &
+unwrapSymbol(const Fortran::parser::AllocateObject &allocObj) {
+  const auto &lastName = Fortran::parser::GetLastName(allocObj);
+  assert(lastName.symbol);
+  return lastName.symbol->GetUltimate();
+}
+
+namespace {
+// Lower ALLOCATE/DEALLOCATE stmt ERROR and STAT variable as well as the source
+// file location to be passed to the runtime.
+struct ErrorManagementValues {
+  void lower(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+             const Fortran::lower::SomeExpr *statExpr,
+             const ::Fortran::lower::SomeExpr *errMsgExpr) {
+    auto builder = converter.getFirOpBuilder();
+    if (statExpr) {
+      TODO("lower stat expr in allocate and deallocate");
+      hasStat =
+          builder.createIntegerConstant(loc, builder.getIntegerType(1), 1);
+    } else {
+      hasStat =
+          builder.createIntegerConstant(loc, builder.getIntegerType(1), 0);
```

Comment:
We could add a `createBool` to the fir builder.

Reply:
I applied this suggestion, that is more readable and sounds useful in other contexts.